util/tracecpu/collector.go

```diff
@@ -0,0 +1,333 @@
+	"time"
+
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"github.com/wangjohn/quickselect"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+// TopSQLRecord represents a single record of how much cpu time a sql plan consumes in one second.
+//
+// PlanDigest can be empty, because:
+// 1. some sql statements has no plan, like `COMMIT`
+// 2. when a sql statement is being compiled, there's no plan yet
+type TopSQLRecord struct {
+	SQLDigest  string
+	PlanDigest string
+	CPUTimeMs  uint32
+}
+
+type planBinaryDecodeFunc func(string) (string, error)
+
+// TopSQLDataPoints represents the cumulative SQL plan CPU time in current minute window
+type TopSQLDataPoints struct {
+	CPUTimeMsList  []uint32
+	TimestampList  []uint64
+	CPUTimeMsTotal uint64
+}
+
+type digestAndCPUTime struct {
+	Key            string
+	CPUTimeMsTotal uint64
+}
+type digestAndCPUTimeSlice []digestAndCPUTime
+
+func (t digestAndCPUTimeSlice) Len() int {
+	return len(t)
+}
+
+// We need find the kth largest value, so here should use >
+func (t digestAndCPUTimeSlice) Less(i, j int) bool {
+	return t[i].CPUTimeMsTotal > t[j].CPUTimeMsTotal
+}
+func (t digestAndCPUTimeSlice) Swap(i, j int) {
+	t[i], t[j] = t[j], t[i]
+}
+
+type planRegisterJob struct {
+	planDigest     string
+	normalizedPlan string
+}
+
+// TopSQLCollector is called periodically to collect TopSQL resource usage metrics
+type TopSQLCollector struct {
+	// // calling this can take a while, so should not block critical paths
+	// planBinaryDecoder planBinaryDecodeFunc
+	mu sync.RWMutex
+
+	// topSQLMap maps `sqlDigest-planDigest` to TopSQLDataPoints
+	topSQLMap map[string]*TopSQLDataPoints
+	maxSQLNum int
+
+	// normalizedSQLMap is an map, whose keys are SQL digest strings and values are normalized SQL strings
+	normalizedSQLMap map[string]string
+
+	// normalizedPlanMap is a plan version of normalizedSQLMap
+	// this should only be set from the dedicated worker
+	normalizedPlanMap map[string]string
+
+	planRegisterChan chan *planRegisterJob
+
+	// current tidb-server instance ID
+	instanceID string
+
+	agentGRPCAddress string
+
+	quit chan struct{}
+}
+
+// TopSQLCollectorConfig is the config for TopSQLCollector
+type TopSQLCollectorConfig struct {
+	PlanBinaryDecoder   planBinaryDecodeFunc
+	MaxSQLNum           int
+	SendToAgentInterval time.Duration
+	AgentGRPCAddress    string
+	InstanceID          string
+}
+
+func encodeCacheKey(sqlDigest, planDigest string) string {
+	return sqlDigest + "-" + planDigest
+}
+
+func decodeCacheKey(key string) (string, string) {
+	split := strings.Split(key, "-")
+	sqlDigest := split[0]
+	PlanDigest := split[1]
+	return sqlDigest, PlanDigest
+}
+
+func newAgentClient(addr string, sendingTimeout time.Duration) (*grpc.ClientConn, tipb.TopSQLAgent_CollectCPUTimeClient, context.CancelFunc, error) {
+	dialCtx, cancel := context.WithTimeout(context.TODO(), time.Second)
+	defer cancel()
+	conn, err := grpc.DialContext(dialCtx, addr, grpc.WithInsecure())
```

Comment:
![image](https://user-images.githubusercontent.com/1916485/120304065-04a1de00-c302-11eb-9798-1d4c43d9632d.png)

Seems that the default backoff does not suit our demands when remote agent is down, for example, it will maximum not connect to the agent for 120s. I think you need to adjust this config. Current configurations can be a good example.